Merge all features related to C# 7 ref features (#795)
dotnet/csharpstandard: standard/classes.md

- If the declaration includes the `sealed` modifier, then the declaration also includes the `override` modifier.
- If the declaration includes the `partial` modifier, then it does not include any of the following modifiers: new, `public`, `protected`, `internal`, `private`, `virtual`, `sealed`, `override`, `abstract`, or `extern`.

It is a compile-time error to have both `ref` and a *return_type* of `void`.

> If `ref` is present, the method ***returns-by-ref***; otherwise, if *return_type* is `void`, the method ***returns-no-value***; otherwise, the method ***returns-by-value***.

Reviewer: This appears to be defining verbs.  Yet elsewhere these terms are used as adjectives.  Which is it?

Author: adjectives, for consistency.

The formal parameter list consists of one or more comma-separated parameters of which only the last may be a *parameter_array*.

> A *fixed_parameter* consists of an optional set of *attributes* ([§22](attributes.md#22-attributes)); an optional `in`, `out`, `ref` modifier; a *type*; an *identifier*; and an optional *default_argument*. Each *fixed_parameter* declares a parameter of the given type with the given name. The `this` modifier designates the method as an extension method and is only allowed on the first parameter of a static method in a non-generic, non-nested static class. If the parameter is a `struct` type or a type parameter constrained to a `struct`, the `this` modifier may be combined with either the `ref` or `in` modifier, but not the `out` modifier. Extension methods are further described in [§15.6.10](classes.md#15610-extension-methods). A *fixed_parameter* with a *default_argument* is known as an ***optional parameter***, whereas a *fixed_parameter* without a *default_argument* is a ***required parameter***. A required parameter may not appear after an optional parameter in a *formal_parameter_list*.

Reviewer: This first sentence used to permit a `this` modified.  Now it does not.  That needs to be added back in.

A method invocation ([§12.8.9.2](expressions.md#12892-method-invocations)) creates a copy, specific to that invocation, of the formal parameters and local variables of the method, and the argument list of the invocation assigns values or variable references to the newly created formal parameters. Within the *block* of a method, formal parameters can be referenced by their identifiers in *simple_name* expressions ([§12.8.4](expressions.md#1284-simple-names)).

The following kinds of formal parameters exist:

- Value parameters, which are declared without any modifiers.
- Input parameters, which are declared with the `in` modifier.
- Output parameters, which are declared with the `out` modifier.
- Reference parameters, which are declared with the `ref` modifier.
- Parameter arrays, which are declared with the `params` modifier.

> > *Note*: As described in [§7.6](basic-concepts.md#76-signatures-and-overloading), the `in`, `out`, and `ref` modifiers are part of a method's signature, but the `params` modifier is not.

Reviewer 2: Presumably we expect the curly quote to be fixed up by the smart-quotes action? We should make sure we rerun that before shipping v7...

Author: yes, and yes.

A method is permitted to assign new values to a value parameter. Such assignments only affect the local storage location represented by the value parameter—they have no effect on the actual argument given in the method invocation.

> #### §method-input-parameters-new-clause Input parameters

Reviewer 3: Was this meant to be a section number?

Author: It will be. The automatic section renumbering tool runs when we merge a PR into the main branch. It renumbers any sections, and updates all links based on the new numbers. That key lets us add links to new sections in the same PR.

> #### §method-input-parameters-new-clause Input parameters

Reviewer 4: This clause and the following two on reference and output parameters contain some common text which results in triplication. We should consider pulling this out into its own clause, and maybe making these three clauses sub-clauses of this clause.

This clause could be titled “By-ref parameters" following the “returns-by-ref" language used for reference returns, and because "Reference parameters" is already taken!

For the moment I will not attempt to de-triplicate as I'm confusing Github enough as it is~ ;-)

I will make similar suggestions for all three clauses to align them with the intorduction of reference variables (§fix_up_probably_9_7).

Author: This is a great suggestion. I'd like to flag this as a v7 milestone issue and create a separate PR to de-triplicate.

Reviewer 2: Deferring until after this PR is merged, but before C# 7 ships. See #799.

> A parameter declared with an `in` modifier is an input parameter. Unlike a value parameter, an input parameter does not create a new storage location. Instead, an input parameter represents the same storage location as the variable given as the argument, or created by the implementation ([§12.6.2.3](expressions.md#12623-run-time-evaluation-of-argument-lists)), in the method invocation.

Reviewer 4: With the introduction of reference variables we now acknowledge that ref/in/out parameters do create storage locations which contain a referent…

I've included a *Note* that the referent can be changed, which it can be in Roslyn. If this is *__not__* intended that should be stated in normative text.

**Note: I've left the "created by the implementation ([§12.6.2.3](expressions.md#12623-run-time-evaluation-of-argument-lists))" in but cannot locate this clause – I'm guessing it will be added?**

```suggestion
A parameter declared with an `in` modifier is an input parameter. An input parameter corresponds to a local reference variable (§fix_up_probably_9_7) that gets its initial referent from the corresponding argument supplied in the method invocation. That argument is either a variable existing at the point of the method invocation, or one created by the implementation ([§12.6.2.3](expressions.md#12623-run-time-evaluation-of-argument-lists)) in the method invocation.

*Note*: As with reference variables the referent of an input parameter can be changed using the ref assignment (`= ref`) operator, however the value stored in the referent itself cannot be changed.
```

Author: The change for "created by an implementation" is in this PR: https://github.com/dotnet/csharpstandard/pull/795/files#diff-d9d678d0b42d3c668ebd64159dadfd6188361121feca8fb94fddf8816433cd14R603

(starting at line 603 in expressions.md)

Reviewer 4: @BillWagner – Thanks, I just break Github, couldn't find the referenced clause anywhere even using its improved search to look for the title and it showed me nothing, and today there it is and I can hear Github sniggering ;-) Anyway, the x-ref in the suggested change is good :-)

When a formal parameter is an input parameter, the corresponding argument in a method invocation shall consist of either the keyword `in` followed by a *variable_reference* (§input-parameters-new-clause) of the same type as the formal parameter, or an *expression* for which an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) exists from that argument expression to the type of the corresponding parameter. A variable shall be definitely assigned before it can be passed as an input parameter.

It is a compile-time error to modify the value of an input parameter.

Within a method, an input parameter is always considered definitely assigned.

A method declared as an iterator ([§15.14](classes.md#1514-iterators)) may not have input parameters.

In a method that takes input parameters, it is possible for multiple names to represent the same storage location.

#### 15.6.2.3 Reference parameters

> A parameter declared with a `ref` modifier is a reference parameter. Unlike a value parameter, a reference parameter does not create a new storage location. Instead, a reference parameter represents the same storage location as the variable given as the argument in the method invocation.

Reviewer 4: With the introduction of reference variables we now acknowledge that ref/in/out parameters do create storage locations which contain a referent…

I've included a *Note* that the referent can be changed, which it can be in Roslyn. If this is *__not__* intended that should be stated in normative text.

```suggestion
A parameter declared with a `ref` modifier is a reference parameter. A reference parameter corresponds to a local reference variable (§fix_up_probably_9_7) that gets its initial referent from the corresponding argument supplied in the method invocation.

*Note*: As with reference variables the referent of a reference parameter can be changed using the ref assignment (`= ref`) operator.
```

## I've broken Github, again…

Github is refusing to allow me to suggest the corresponding change to the opening paragraph for output parameters:

> #### 15.6.2.4 Output parameters
> A parameter declared with an `out` modifier is an output parameter. Similar to a reference parameter, an output parameter does not create a new storage location. Instead, an output parameter represents the same storage location as the variable given as the argument in the method invocation.

So here it is, just not as a clickable change:

> #### 15.6.2.4 Output parameters
> A parameter declared with an `out` modifier is an output parameter. An output parameter corresponds to a local reference variable (§fix_up_probably_9_7) that gets its initial referent from the corresponding argument supplied in the method invocation.

A method is permitted to assign new values to a value parameter. Such assignments only affect the local storage location represented by the value parameter—they have no effect on the actual argument given in the method invocation.

#### §method-input-parameters-new-clause Input parameters

> A parameter declared with an `in` modifier is an input parameter. An input parameter corresponds to a local reference variable (§ref-span-safety) that gets its initial referent from the corresponding argument supplied in the method invocation. That argument is either a variable existing at the point of the method invocation, or one created by the implementation ([§12.6.2.3](expressions.md#12623-run-time-evaluation-of-argument-lists)) in the method invocation.

Reviewer: This first sentence is confusing.  What does "corresponds to" mean?  That the caller has to pass a local reference variable?  Of course, that's not right, but some readers may be misled.  I was misled on the first reading.

Reviewer 4: This description is following the pattern from above:

> #### 15.6.2.2 Value parameters
>
> A parameter declared with no modifiers is a value parameter. A value parameter corresponds to a local variable that gets its initial value from the corresponding argument supplied in the method invocation.

Personally I’m happy with “corresponds to” but a simple alternative might be “is a”.

My thinking: In the case of input parameters the variable is a "read-only value, read-write referent, local reference variable" (which the following text and note explains) and I read “is a local reference variable” as implying a stronger equivalence than wanted here while “corresponds to a local reference variable“ does not, hence my “might be”…

English is hard ;-)

Reviewer 2: Discussion result: change "corresponds to" to "is a" on each occurrence (value, ref, in, out), within this PR.

*Note*: As with reference variables the referent of an input parameter can be changed using the ref assignment (`= ref`) operator, however the value stored in the referent itself cannot be changed.

When a formal parameter is an input parameter, the corresponding argument in a method invocation shall consist of either the keyword `in` followed by a *variable_reference* (§input-parameters-new-clause) of the same type as the formal parameter, or an *expression* for which an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) exists from that argument expression to the type of the corresponding parameter. A variable shall be definitely assigned before it can be passed as an input parameter.

It is a compile-time error to modify the value of an input parameter.

Within a method, an input parameter is always considered definitely assigned.

> A method declared as an iterator ([§15.14](classes.md#1514-iterators)) may not have input parameters.

Reviewer 2: Mention async methods here as well? (And for ref/out - or in the detriplicated version, anyway.)

In a method that takes input parameters, it is possible for multiple names to represent the same storage location.

#### 15.6.2.3 Reference parameters

A parameter declared with a `ref` modifier is a reference parameter. A reference parameter corresponds to a local reference variable (§ref-span-safety) that gets its initial referent from the corresponding argument supplied in the method invocation.

> 

Reviewer: "corresponds" is confusing here.  It is "like"?

Reviewer 4: See comment on *Input Parameters* above – change all or none.

Reviewer 2: This is covered by the above - same resolution, fix in this PR.

> *Example*: The types `string[]` and `string[][]` can be used as the type of a parameter array, but the type `string[,]` can not. *end example*

> It is not possible to combine the `params` modifier with the modifiers `in`, `out`, and `ref`.

Reviewer: Do we mean it is an error if you do?

Reviewer 4: That’s been there a very long time, no idea why this form of language was used. Line 1971 above:

> It is a compile-time error to have both `ref` and a *return_type* of `void`.

suggests the wording:

```suggestion
It is a compile-time error to combine the `params` modifier with the modifiers `in`, `out`, or `ref`.
```

(I changed “and” to “or” as it might be read that `params in out ref` is invalid?)

Note however that the grammar does not allow `params` and one of the *parameter_mode_modifier* anyway so an alternative suggestion is:

```suggestion
```

Or make it a note:

```suggestion
*Note*: It is not possible to combine the `params` modifier with the modifiers `in`, `out`, or `ref`. *end note*
```

Choice, choices, take your pick! 🙂

Author: I prefer the third option.

### 15.6.10 Extension methods

When the first parameter of a method includes the `this` modifier, that method is said to be an ***extension method***. Extension methods shall only be declared in non-generic, non-nested static classes. The first parameter of an extension method is restricted, as follows:

> - It may have the parameter modifier `in` provided the parameter has a value type

Reviewer: "May" does not seem like a restriction.  Perhaps "provided" is intended to say "only if"?
May the `out` modifier be used?

Reviewer 4: “only if” – 👍🏼

Author: Changed both this and the following bullet to "only if"

- It may have the parameter modifier `ref` provided the parameter has a value type or is a generic type constrained to struct

> - It may not be a pointer type.

Reviewer 2: Shall rather than may? ("May not" *always* feels ambiguous to me, between "must not" and "need not". This case is "must not" but if I were to say "I may not attend the next meeting" that would be very different...) Arguably it's worth a trawl through the whole standard for "may not" at some point.

Author: Fixed here, but not throughout the standard

When the effective return type of a method is `void` and the method has a block body, `return` statements ([§13.10.5](statements.md#13105-the-return-statement)) in the block shall not specify an expression. If execution of the block of a void method completes normally (that is, control flows off the end of the method body), that method simply returns to its caller.

When the effective return type of a method is `void` and the method has an expression body, the expression `E` shall be a *statement_expression*, and the body is exactly equivalent to a block body of the form `{ E; }`.

For a returns-by-value method ([§15.6.1](classes.md#1561-general)), each return statement in that method's body shall specify an expression that is implicitly convertible to the effective return type.

For a returns-by-ref method ([§15.6.1](classes.md#1561-general)), each return statement in that method's body shall specify an expression whose type is that of the effective return type, and has a *ref_safe_scope* of *calling_method* (§ref-safe-contexts).

> The endpoint of the method body shall not be reachable. In other words, control is not permitted to flow off the end of the method body.

Reviewer: That forbids

```csharp
void M() {
}
```

Reviewer 4: It does indeed. In keeping with the *returns-by-xxx* changes I suggest:

```suggestion
For returns-by-value and returns-by-ref methods the endpoint of the method body shall not be reachable. In other words, control is not permitted to flow off the end of the method body.
```

Property declarations are subject to the same rules as method declarations ([§15.6](classes.md#156-methods)) with regard to valid combinations of modifiers.

> The *type* of a property declaration specifies the type of the property introduced by the declaration. If the `ref` modifier is present, the return and the `set` parameter are a *variable_reference* (§9.4), that is optionally read-only.The *member_name* ([§15.6.1](classes.md#1561-general)) specifies the name of the property. Unless the property is an explicit interface member implementation, the *member_name* is simply an *identifier*. For an explicit interface member implementation ([§18.6.2](interfaces.md#1862-explicit-interface-member-implementations)), the *member_name* consists of an *interface_type* followed by a “`.`” and an *identifier*.

Reviewer 5: >and the `set` parameter [](http://example.com/codeflow?start=148&length=23)

Is the "`set` parameter" related to the `set` accessor? If so, does this apply since ref-returning properties cannot have `set` accessors?

> The *type* of a property declaration specifies the type of the property introduced by the declaration. If the `ref` modifier is present, the return and the `set` parameter are a *variable_reference* (§9.4), that is optionally read-only.The *member_name* ([§15.6.1](classes.md#1561-general)) specifies the name of the property. Unless the property is an explicit interface member implementation, the *member_name* is simply an *identifier*. For an explicit interface member implementation ([§18.6.2](interfaces.md#1862-explicit-interface-member-implementations)), the *member_name* consists of an *interface_type* followed by a “`.`” and an *identifier*.

Reviewer 5: >read-only.The [](http://example.com/codeflow?start=226&length=13)

Missing space.

Property declarations are subject to the same rules as method declarations ([§15.6](classes.md#156-methods)) with regard to valid combinations of modifiers.

> The *type* of a property declaration specifies the type of the property introduced by the declaration. If the `ref` modifier is present, the return parameter is a *variable_reference* (§9.4), that is optionally read-only. The *member_name* ([§15.6.1](classes.md#1561-general)) specifies the name of the property. Unless the property is an explicit interface member implementation, the *member_name* is simply an *identifier*. For an explicit interface member implementation ([§18.6.2](interfaces.md#1862-explicit-interface-member-implementations)), the *member_name* consists of an *interface_type* followed by a “`.`” and an *identifier*.

Reviewer 2: Do we ever define "return parameter"? I don't remember seeing that term before.

Author: The term "expression of a return statement" is used in 13.10.5. Edit to make it consistent here.

A *property_initializer* may only be given for an automatically implemented property ([§15.7.4](classes.md#1574-automatically-implemented-properties)), and causes the initialization of the underlying field of such properties with the value given by the *expression*.

> Even though the syntax for accessing a property is the same as that for a field, a property is not classified as a variable. Thus, it is not possible to pass a property as a an `in`, `out`, or `ref` argument. However, a property can be passed as an *expression* to a method having an input parameter.

Reviewer 5: >as a an [](http://example.com/codeflow?start=169&length=7)

Typo: "as an"

A *property_initializer* may only be given for an automatically implemented property ([§15.7.4](classes.md#1574-automatically-implemented-properties)), and causes the initialization of the underlying field of such properties with the value given by the *expression*.

> Even though the syntax for accessing a property is the same as that for a field, a property is not classified as a variable. Thus, it is not possible to pass a property as an `in`, `out`, or `ref` argument. However, a property can be passed as an *expression* to a method having an input parameter.

Reviewer: Correction: A `ref` property is classified as a variable.

Reviewer 4: Yes and no… A property is not classified as a variable but a property may (now) return a *variable_reference*. This same issue was spotted by @cston for indexers and I think we can borrow the fix from there:

```suggestion
*Note*: Even though the syntax for accessing a property is the same as that for a field, a property is not classified as a variable. Thus, it is not possible to pass a property as an `in`, `out`, or `ref` argument unless the property itself returns a reference (§ref-span-safety). *end note*
```

Reviewer 4: Update: It is possible that ref returning properties are not in v7, see indexer comments below.

Author: `ref` properties are allowed in V7. I recommend accepting Nigel's suggestion.

Reviewer 2: We believe Nigel's suggestion is correct. Committing his suggestion - @gafter please reopen if you disagree.

> *Note:* As indexers are designed to be used in array element-like contexts, the term *element type* as defined for an array is also used with an indexer. *end note*

> The *formal_parameter_list* specifies the parameters of the indexer. The formal parameter list of an indexer corresponds to that of a method ([§15.6.2](classes.md#1562-method-parameters)), except that at least one parameter shall be specified, and that the `this`, `in`, `out`, and `ref` parameter modifiers are not permitted.

Reviewer 5: >`in` [](http://example.com/codeflow?start=265&length=4)

Aren't `in` parameters allowed? `object this[in int index] { get; }`

Reviewer 4: @cston – Yes, they are, thanks for spotting this. Would you like to enter a suggested fix (using the +/- page icon in the editing bar just above a comment)?

Now I wonder where else `in` has been added by mistake…

Author: updated in upcoming commit.

An expression body consisting of “`=>`” followed by an expression `E` and a semicolon is exactly equivalent to the block body `{ get { return E; } }`, and can therefore only be used to specify read-only indexers where the result of the get accessor is given by a single expression.

> Even though the syntax for accessing an indexer element is the same as that for an array element, an indexer element is not classified as a variable. Thus, it is not possible to pass an indexer element as an `in`, `out`, or `ref` argument.

Reviewer 5: >Thus, it is not possible to pass an indexer element as an `in`, `out`, or `ref` argument. [](http://example.com/codeflow?start=150&length=89)

A ref-returning indexer element can be passed by ref.

Reviewer 4: @cston – Yes, it certainly can. I'm going to suggest:

```suggestion
*Note*: Even though the syntax for accessing an indexer element is the same as that for an array element, an indexer element is not classified as a variable. Thus, it is not possible to pass an indexer element as an `in`, `out`, or `ref` argument unless the indexer itself returns a reference (§FIX_UP_probably_9_7).
```

Making it a *Note*, as it really always was.

I suspect this is an issue stemming from how reference returns impact little corners of the Standard…

An expression body consisting of “`=>`” followed by an expression `E` and a semicolon is exactly equivalent to the block body `{ get { return E; } }`, and can therefore only be used to specify read-only indexers where the result of the get accessor is given by a single expression.

> *Note*: Even though the syntax for accessing an indexer element is the same as that for an array element, an indexer element is not classified as a variable. Thus, it is not possible to pass an indexer element as an `in`, `out`, or `ref` argument unless the indexer itself returns a reference (§ref-span-safety).

Reviewer 2: Assuming we really want this to change from normative to informative (which is probably fine), we should have an *end note* at the end of it.

> > *Note*: Even though the syntax for accessing an indexer element is the same as that for an array element, an indexer element is not classified as a variable. Thus, it is not possible to pass an indexer element as an `in`, `out`, or `ref` argument unless the indexer itself returns a reference (§ref-span-safety). *end note*

Reviewer: The syntax above does not permit declaraing an indexer to return a ref.

Reviewer 4: It does not…

Hmm, is the syntax behind (pre v7) or the note ahead (post v7)?

If the former then the correct syntax would be:

```ANTLR
indexer_declarator
    : ('ref' 'readonly'?)? type 'this' '[' formal_parameter_list ']'
    | ('ref' 'readonly'?)? type interface_type '.' 'this' '[' formal_parameter_list ']'
    ;
```

Update: A quick look seems to show that ref returning indexers and properties are **not** in v7.3 but are in v9 (no compiler to hand for versions between those).

Author: Verified that indexers and properties can return by-ref in 7.3. This compiles in C# 7.3:

```csharp
    public class C
    {
        private int number;

        private int[] storage = new int[20];
        public ref int InternalRefReturn
        {
            get { return ref number; }
        }

        public ref int this[int index]
        {
            get { return ref storage[index]; }
        }
    }
```

Reviewer 2: @BillWagner to pick up Nigel's grammar change in this PR.

An iterator block may be used as the body of a function member as long as the return type of the corresponding function member is one of the enumerator interfaces ([§15.14.2](classes.md#15142-enumerator-interfaces)) or one of the enumerable interfaces ([§15.14.3](classes.md#15143-enumerable-interfaces)). It may occur as a *method_body*, *operator_body* or *accessor_body*, whereas events, instance constructors, static constructors and finalizers may not be implemented as iterators.

> When a function member is implemented using an iterator block, it is a compile-time error for the formal parameter list of the function member to specify any `in`, `out`, or `ref` parameters.

Reviewer: What about parameters of a ref struct type?

Author: Iterators may have parameters of a `ref` struct type.

Reviewer 2: We're not sure if this will work - looks like it compiles.

```csharp
public IEnumerable<string> Top(Span<byte> span)
{
    yield return "1";
    Console.WriteLine(span[0]);
    yield return "2";
}

public IEnumerable<string> Middle()
{
     Span<byte> span = stackalloc byte[5];
     IEnumerable<string> ret = Top(span);
     return ret;
}

public void Bottom()
{
    foreach (var value in Middle())
    {
    }
}
```

Ah, no it doesn't:  CS4013 Instance of type `'Span<byte>'` cannot be used inside a nested function, query expression, iterator block or async method

@BillWagner: Neal's original comment about ref struct types is correct, and should be included.

A method ([§15.6](classes.md#156-methods)) or anonymous function ([§12.19](expressions.md#1219-anonymous-function-expressions)) with the `async` modifier is called an ***async function***. In general, the term ***async*** is used to describe any kind of function that has the `async` modifier.

> It is a compile-time error for the formal parameter list of an async function to specify any `in`, `out`, or `ref` parameters.

Reviewer: What about parameters of a ref struct type?
